core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,355 @@
+        CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES_DEFAULT);
+
+    return enabled && durationMs >= 0 && totalBytes > 0;
+  }
+
+  private ContentCache fileContentCache;
+  private Map<String, String> properties;
+  private FileIO wrappedIO;
+
+  /**
+   * CachingFileIO is not meant to be dynamically loaded.
+   */
+  private CachingFileIO() {
+  }
+
+  public CachingFileIO(FileIO io) {
+    assignIO(io);
+  }
+
+  public void assignIO(FileIO io) {
+    this.wrappedIO = io;
+  }
+
+  /**
+   * Initialize File IO from catalog properties.
+   * <p>
+   * It assumes that the inner FileIO is already initialized and therefore does not reinitialize it.
+   * If this is a reinitialization, it is the caller responsibility to call {@link #close()} and
+   * {@link #assignIO(FileIO)} prior to reinitialization to close the previously assigned FileIO and reassign a new one.
+   * Otherwise, previously assigned FileIO will be kept in use.
+   * <p>
+   * Reinitialization will discard previous content cache unless shared cache is being used
+   * ({@link CatalogProperties#IO_CACHE_SHARED} stays true).
+   * @param newProperties catalog properties
+   */
+  @Override
+  public void initialize(Map<String, String> newProperties) {
+    this.properties = newProperties;
+    initCache();
+  }
+
+  private void initCache() {
+    boolean isSharedCache = PropertyUtil.propertyAsBoolean(properties,
+        CatalogProperties.IO_CACHE_SHARED,
+        CatalogProperties.IO_CACHE_SHARED_DEFAULT);
+    if (isSharedCache) {
+      this.fileContentCache = getOrCreate(properties);
+      LOG.info("CachingFileIO created with shared cache. {} ", this.fileContentCache);
+    } else {
+      this.fileContentCache = createCache(properties);
+      LOG.info("CachingFileIO created. {} ", this.fileContentCache);
+    }
+  }
+
+  @Override
+  public void close() {
+    if (wrappedIO != null) {
+      wrappedIO.close();
+      wrappedIO = null;
+    }
+  }
+
+  @Override
+  public InputFile newInputFile(String path) {
+    InputFile inFile = wrappedIO.newInputFile(path);
+    if (fileContentCache != null && path.contains("/metadata/")) {
+      // TODO: Currently we just believe that this is a metadata file if it has "metadata" dir on its path. But metadata
+      //  location can be set differently for different tables. A table might have different metadata path by
+      //  setting custom "write.metadata.path" TableProperties.
+      //  We might need to extend FileIO interface. Add method 'newMetadataInputFile' and let caller use it to
+      //  differentiate metadata files vs other files.
+      return new CachingInputFile(fileContentCache, inFile);
+    }
+    return inFile;
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return wrappedIO.newOutputFile(path);
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    wrappedIO.deleteFile(path);
+    if (fileContentCache != null) {
+      fileContentCache.cache.invalidate(path);
```

Comment:
If this is going to reach into `fileContentCache` to call methods on `cache` directly, why have `fileContentCache`? Just to check whether it is null? If it is a useful class, I would expect this to call its API instead of modifying its internal state.

Reply:
Added more API method to avoid accessing internal `cache`.